### PR TITLE
RF-GFN

### PR DESCRIPTION
**Pull request by michalkoziarski**

### CODE REVIEW

File: tests/gflownet/envs/common.py [hunk 1]
```diff
@@ -204,7 +204,15 @@ def test__get_parents__returns_same_state_and_eos_if_done(env):
...
-        assert all([torch.equal(p, env.state) for p in parents])
+        assert all(
+            [torch.equal(torch.isnan(p), torch.isnan(env.state)) for p in parents]
+        )
+        assert all(
+            [
+                torch.equal(p[~torch.isnan(p)], env.state[~torch.isnan(env.state)])
+                for p in parents
+            ]
+        )
```
Comment (alexhernandezgarcia) [inline tests/gflownet/envs/common.py:215]: Why was this needed?

Comment (alexhernandezgarcia) [reply tests/gflownet/envs/common.py:215]: Ah it's because the states of the tree are initialised to `torch.nan`?

Comment (michalkoziarski) [reply tests/gflownet/envs/common.py:215]: Yes, and (get ready to be amazed):
```
>>> torch.nan == torch.nan
False
```

Comment (alexhernandezgarcia) [reply tests/gflownet/envs/common.py:215]: LOL

Comment (alexhernandezgarcia) [reply tests/gflownet/envs/common.py:215]: I guess it makes sense (ish) if I think about it... At least it's consistent with numpy:
```
>>> numpy.inf * 0 == numpy.inf * 0
False
```

Comment (michalkoziarski) [reply tests/gflownet/envs/common.py:215]: Well riddle me this:
```
>>> torch.inf == torch.inf
True
```
and
```
>>> torch.nan in [torch.nan]
True
```

Comment (alexhernandezgarcia) [reply tests/gflownet/envs/common.py:215]: hahaha I can see the logic of either way...

File: gflownet/envs/tree.py [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+    def step(
+        self, action: Tuple[int, int, float]
+    ) -> Tuple[List[int], Tuple[int, int, float], bool]:
+        # If action not found in action space raise an error.
+        if action[:2] not in self.action_space:
+            raise ValueError(
+                f"Tried to execute action {action} not present in action space."
+            )
+        else:
+            action_idx = self.action_space.index(action[:2])
```
Comment (alexhernandezgarcia) [inline gflownet/envs/tree.py:355]: Why `action[:2]` instead of just `action`?

Comment (alexhernandezgarcia) [reply gflownet/envs/tree.py:355]: Ok, got it by myself - actions are tuples of three things. I got confused by the dosctring in the action space method. I am updating it.

File: tests/gflownet/envs/common.py (hunk 1 above)
Comment (alexhernandezgarcia) [reply tests/gflownet/envs/common.py:199]: I have incorporated the check of NaNs in `env.equal(x, y)` so it's easier to check if two states are the same also with nans. See [here](https://github.com/alexhernandezgarcia/gflownet/commit/c347096a001a36858d99b13f6c3223b66ddff2b0)

**Comment by alexhernandezgarcia** [comment]
Test passed after merging locally :+1: Will merge :)

**Review by alexhernandezgarcia — APPROVED** [review]
Looking goooood!
